mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -443,6 +444,144 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
+
+  LogicalResult
+  matchAndRewrite(ArithOp op, typename ArithOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type type = this->getTypeConverter()->convertType(op.getType());
+    if (!isa_and_nonnull<IntegerType, emitc::SignedSizeTType, emitc::SizeTType>(
+            type)) {
+      return rewriter.notifyMatchFailure(
+          op, "expected integer or size_t/ssize_t type");
+    }
+
+    if (type.isInteger(1)) {
+      return rewriter.notifyMatchFailure(op, "i1 type is not implemented");
+    }
+
+    Type arithmeticType = adaptIntegralTypeSignedness(type, isUnsignedOp);
+
+    Value lhs = adaptValueType(adaptor.getLhs(), rewriter, arithmeticType);
+    Value rhs = adaptValueType(adaptor.getRhs(), rewriter, arithmeticType);
+
+    // Add a runtime check for overflow
+    // This is an abuse of the size_t type since we're potentially using values
+    // below -1.
+    Value width;
+    Type sizeTType = (isa<IntegerType>(type)) ? arithmeticType
+                     : (isUnsignedOp)
+                         ? (Type)(emitc::SizeTType::get(op.getContext()))
+                         : (Type)(emitc::SignedSizeTType::get(op.getContext()));
+    if (isa<emitc::SignedSizeTType, emitc::SizeTType>(type)) {
+      Value eight = rewriter.create<emitc::ConstantOp>(
+          op.getLoc(), sizeTType, rewriter.getIndexAttr(8));
+      emitc::CallOpaqueOp sizeOfCall = rewriter.create<emitc::CallOpaqueOp>(
+          op.getLoc(), sizeTType, "sizeof", SmallVector<Value, 1>({eight}));
+      width = rewriter.create<emitc::MulOp>(op.getLoc(), sizeTType, eight,
+                                            sizeOfCall.getResult(0));
+    } else {
+      width = rewriter.create<emitc::ConstantOp>(
+          op.getLoc(), sizeTType,
+          rewriter.getIntegerAttr(sizeTType, type.getIntOrFloatBitWidth()));
+    }
+
+    Value oobTest;
+    Value excessCheck = rewriter.create<emitc::CmpOp>(
+        op.getLoc(), rewriter.getI1Type(), emitc::CmpPredicate::lt, rhs, width);
+    oobTest = excessCheck;
+    if (!isUnsignedOp) {
+      Value zero = rewriter.create<emitc::ConstantOp>(
+          op.getLoc(), sizeTType,
+          (isa<IntegerType>(sizeTType) ? rewriter.getIntegerAttr(sizeTType, 0)
+                                       : rewriter.getIndexAttr(0)));
+      Value defaultCheck =
+          rewriter.create<emitc::CmpOp>(op.getLoc(), rewriter.getI1Type(),
+                                        emitc::CmpPredicate::ge, rhs, zero);
+      oobTest = rewriter.create<emitc::LogicalAndOp>(
+          op.getLoc(), rewriter.getI1Type(), excessCheck, defaultCheck);
+    }
+
+    Value poison = rewriter.create<emitc::ConstantOp>(
+        op.getLoc(), arithmeticType,
+        (isa<IntegerType>(arithmeticType)
+             ? rewriter.getIntegerAttr(arithmeticType, 0)
+             : rewriter.getIndexAttr(0)));
+
+    emitc::ExpressionOp ternary = rewriter.create<emitc::ExpressionOp>(
```

Comment:
Nice use of `emitc::ExpressionOp`. What does the generated C++ code look like?

Reply:
The shift operation (not just its result) must be in the same expression as the ternary op to benefit from the conditional evaluation and avoid UB, hence this use of `ExpressionOp`.

```C++
void shift() {
  // Constants and their signed casts
  size_t v1 = 160;
  size_t v2 = 64;
  ssize_t v3 = (ssize_t) v1;
  ssize_t v4 = (ssize_t) v2;

  // Here goes the right shift op
  // Predicate : (shift amount < 8*sizeof(ssize_t)) && (shift amount >= 0)
  ssize_t v5 = 8;
  ssize_t v6 = sizeof(v5);
  ssize_t v7 = v5 * v6;
  bool v8 = v4 < v7;
  ssize_t v9 = 0;
  bool v10 = v4 >= v9;
  bool v11 = v8 && v10;

  ssize_t v12 = 0; // Poison value
  // Actual shift op.
  size_t v13 = (size_t) v11 ? v3 >> v4 : v12;
  return;
}
```

Reply:
Looks good! On the last line `size_t v13 = (size_t) v11 ? v3 >> v4 : v12;`, why is there a cast? `v11` is already bool and that is the type the conditional parts needs to have.

Reply:
Per [this page](https://en.cppreference.com/w/cpp/language/operator_precedence), the cast is on `v11` indeed, where in the MLIR it is on the result of the expression. That looks a bug in the C++ emission phase.

The EmitC MLIR
```mlir
    %8 = emitc.expression : !emitc.ssize_t {
      %10 = emitc.bitwise_right_shift %2, %1 : (!emitc.ssize_t, !emitc.size_t) -> !emitc.ssize_t
      %11 = emitc.conditional %6, %10, %7 : !emitc.ssize_t
      emitc.yield %11 : !emitc.ssize_t
    }
    %9 = emitc.cast %8 : !emitc.ssize_t to !emitc.size_t
```

is turned into this C++
```c++
size_t v9 = (size_t) v7 ? v3 >> v2 : v8;
```

so it looks like we're missing some parentheses.

Reply:
[Raised an issue](https://github.com/llvm/llvm-project/issues/93470) on upstream LLVM.